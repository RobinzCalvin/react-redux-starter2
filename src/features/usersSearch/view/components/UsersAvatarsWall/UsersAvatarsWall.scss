.users-avatars-wall {
  display: flex;
  justify-content: center;

  &__avatars {
    position: relative;
    display: flex;
    flex-wrap: wrap;
    align-content: flex-start;
    justify-content: center;
    margin: 0;
    padding: 0;
    list-style: none;

    &_size {
      &_big {
        max-width: 90rem;
      }

      &_small {
        max-width: 45rem;
      }
    }
  }

  &__avatar {
    margin: 0.25rem;
    width: 4rem;
    height: 4rem;
  }

  &__avatar-button {
    padding: 0;
    width: 100%;
    height: 100%;
    outline: none;
    border: none;
    background: none;
    cursor: pointer;
    transition: transform 0.15s;

    &:focus,
    &:hover {
      transform: scale(1.15);
    }
  }

  &__image {
    width: 100%;
    height: 100%;
  }
}
