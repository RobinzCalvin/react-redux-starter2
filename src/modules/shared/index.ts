export { default as Layout } from './Layout/Layout';
export { default as withLayout } from './withLayout';
