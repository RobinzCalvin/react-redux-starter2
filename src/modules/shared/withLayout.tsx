import React from 'react';

import Layout from './Layout/Layout';

function withLayout(title: string) {
  return (Component: React.ComponentType) => () => (
    <Layout title={title}>
      <Component />
    </Layout>
  );
}

export default withLayout;
