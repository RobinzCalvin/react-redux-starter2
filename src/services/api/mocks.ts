import { ICakePreview } from 'shared/types/models';

// tslint:disable:max-line-length
export const cakesPreview: ICakePreview[] =  [
  {
    name: 'Vanilla Almond',
    description: 'Light and sweet but moist and rich in flavor, our Vanilla Almond cake is filled and frosted with our house vanilla buttercream. Vanilla Almond Amycakes are decorated with sparkly clear crystal sprinkles, pearl sprinkles and a ring of buttercream rosettes',
    imageURL: '',
  },
  {
    name: 'Golden Vanilla',
    description: 'Old fashioned, buttery golden yellow cake, just as moist and dense as Grandma used to make. Filled and frosted with our house vanilla buttercream. Golden Vanilla Amycakes are decorated with ascending gold sprinkles, white ganache drizzle and white chocolate curls and flakes',
    imageURL: '',
  },
  {
    name: 'Golden Chocolate',
    description: 'Old fashioned, buttery golden yellow cake, just as moist and dense as Grandma used to make. Filled and frosted with our house chocolate buttercream. Golden Chocolate Amycakes are decorated with ascending gold sprinkles, chocolate ganache drizzle and dark chocolate curls and flakes.',
    imageURL: '',
  },
  {
    name: 'Confetti',
    description: 'Moist Vanilla Almond cake studded with confetti sprinkles in the batter. Confetti Amycakes are frosted with vanilla buttercream and decorated with colorful sprinkles.',
    imageURL: '',
  },
  {
    name: 'Zebra',
    description: 'Alternating layers of vanilla almond and chocolate cakes, filled and frosted with our house vanilla buttercream.  Zebra Amycakes are decorated with chocolate ganache drizzle and white and chocolate curls. ',
    imageURL: '',
  },
];
