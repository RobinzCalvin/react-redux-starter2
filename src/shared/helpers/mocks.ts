import { IProfile, IGithubUser, IDetailedGithubUser, IRepository } from 'shared/types/models';

export function createContainerMock() {
  return function ContainerMock() {
    return null;
  } as any;
}

export function getProfileMock(): IProfile {
  return {
    age: 100,
    nickname: 'Nickname',
    name: 'Name name',
    bio: 'profile bio',
    avatarURL: 'https://the-url.com',
  };
}

export function getGithubUserMock(): IGithubUser {
  return {
    id: 1,
    username: 'username',
    avatarURL: 'http://the-url.com',
    htmlURL: 'http://the-url.com',
  };
}

export function getDetailedGithubUserMock(): IDetailedGithubUser {
  return {
    ...getGithubUserMock(),
    followersNumber: 12,
    followingNumber: 12,
    reposNumber: 12,
    realName: 'The real name',
    location: 'Nowhere',
  };
}

export function getRepositoryMock(): IRepository {
  return {
    id: 1,
    openIssuesNumber: 1,
    starsNumber: 1,
    forksNumber: 1,
    description: 'description',
    name: 'name',
    htmlURL: 'http://the-url.com',
    language: 'en',
    updatedAt: '05/05/2005',
    owner: getGithubUserMock(),
  };
}
// TODO: моки?
