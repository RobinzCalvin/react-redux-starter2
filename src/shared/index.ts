const testContext = (require as any).context('./', true, /^.*\/(?!.*test).*\.tsx?$/);
testContext.keys().forEach(testContext);
