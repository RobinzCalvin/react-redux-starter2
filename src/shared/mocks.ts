import { IProfile, IGithubUser, IDetailedGithubUser, IRepository } from 'shared/types/models';

export const Container: any = () => null;

export const profile: IProfile = {
  age: 100,
  nickname: 'Nickname',
  name: 'Name name',
  bio: 'profile bio',
  avatarURL: 'https://the-url.com',
};

export const githubUser: IGithubUser = {
  id: 1,
  username: 'username',
  avatarURL: 'http://the-url.com',
  htmlURL: 'http://the-url.com',
};

export const detailedGithubUser: IDetailedGithubUser = {
  ...githubUser,
  followersNumber: 12,
  followingNumber: 12,
  reposNumber: 12,
  realName: 'The real name',
  location: 'Nowhere',
};

export const repository: IRepository = {
  id: 1,
  openIssuesNumber: 1,
  starsNumber: 1,
  forksNumber: 1,
  description: 'description',
  name: 'name',
  htmlURL: 'http://the-url.com',
  language: 'en',
  updatedAt: '05/05/2005',
  owner: githubUser,
};
