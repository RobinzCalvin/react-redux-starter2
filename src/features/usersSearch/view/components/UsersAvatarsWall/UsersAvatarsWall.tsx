import React from 'react';
import { autobind } from 'core-decorators';
import block from 'bem-cn';
import * as R from 'ramda';

import { IGithubUser } from 'shared/types/models';
import { Preloader } from 'shared/view/elements';

import { injectSizeToAvatarURL } from '../../../helpers';
import './UsersAvatarsWall.scss';

interface IState {
  areAllAvatarsLoaded: boolean;
}

interface IProps {
  users: IGithubUser[];
  onAvatarClick(user: IGithubUser): void;
}

const b = block('users-avatars-wall');

class UsersAvatarsWall extends React.PureComponent<IProps> {
  public state: IState = {
    areAllAvatarsLoaded: false,
  };

  private avatarSize = 70;

  // { avatarURL: true/false };
  private avatarsLoadingStatus: Record<string, boolean> = (() => {
    const { users } = this.props;
    const avatarsURLs = users.map(x => x.avatarURL);
    return R.zipObj(avatarsURLs, Array(users.length).fill(false));
  })();

  private get wallSize() {
    const { users } = this.props;
    const usersNumber = users.length;
    return usersNumber <= 50
      ? 'small'
      : 'big';
  }

  private get areAllAvatarsLoaded() {
    return Object.values(this.avatarsLoadingStatus).every(R.equals(true));
  }

  public componentDidUpdate({ users: prevUsers }: IProps) {
    const { users } = this.props;
    if (users !== prevUsers) {
      const currentAvatarsURLs = users.map(x => x.avatarURL);
      const prevAvatarsURLs = prevUsers.map(x => x.avatarURL);

      const addedAvatarsURLs = R.difference(currentAvatarsURLs, prevAvatarsURLs);
      const existingAvatarsURLs = R.without(addedAvatarsURLs, currentAvatarsURLs);

      this.avatarsLoadingStatus = {
        ...R.zipObj(
          existingAvatarsURLs,
          existingAvatarsURLs.map(x => this.avatarsLoadingStatus[x]),
        ),
        ...R.zipObj(addedAvatarsURLs, Array(addedAvatarsURLs.length).fill(false)),
      };

      if (!this.areAllAvatarsLoaded) {
        // eslint-disable-next-line react/no-did-update-set-state
        this.setState({ areAllAvatarsLoaded: false });
      }
    }
  }

  public render() {
    const { users } = this.props;
    const { areAllAvatarsLoaded } = this.state;
    return users.length > 0 && (
      <div className={b()}>
        <ul className={b('avatars', { size: this.wallSize })}>
          <Preloader isShown={!areAllAvatarsLoaded} size={100} backgroundColor="#fff" />
          {users.map(this.renderUserAvatar)}
        </ul>
      </div>
    );
  }

  @autobind
  private renderUserAvatar(user: IGithubUser) {
    const { avatarURL } = user;

    return (
      <li key={user.id} className={b('avatar')}>
        <button
          type="button"
          className={b('avatar-button')}
          onClick={this.makeAvatarClickHandler(user)}
          onKeyPress={this.makeAvatarKeyPressHandler(user)}
        >
          <img
            className={b('image')}
            src={injectSizeToAvatarURL(avatarURL, this.avatarSize)}
            alt="Avatar"
            onLoad={this.makeImageOnLoadHandler(user.avatarURL)}
          />
        </button>
      </li>
    );
  }

  @autobind
  private makeAvatarClickHandler(user: IGithubUser) {
    const { onAvatarClick } = this.props;
    return () => onAvatarClick(user);
  }

  @autobind
  private makeAvatarKeyPressHandler(user: IGithubUser) {
    const { onAvatarClick } = this.props;
    return (e: React.KeyboardEvent<HTMLButtonElement>) => e.key === 'Enter' && onAvatarClick(user);
  }

  @autobind
  private makeImageOnLoadHandler(avatarURL: string) {
    return () => {
      this.avatarsLoadingStatus = { ...this.avatarsLoadingStatus, [avatarURL]: true };

      if (this.areAllAvatarsLoaded) {
        this.setState({ areAllAvatarsLoaded: true });
      }
    };
  }
}

export { UsersAvatarsWall, IProps as IUsersAvatarsWallProps };
