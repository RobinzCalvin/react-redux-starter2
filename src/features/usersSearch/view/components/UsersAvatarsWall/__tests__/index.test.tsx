import React from 'react';
import { shallow } from 'enzyme';

import { githubUser } from 'shared/mocks';

import UsersAvatarsWall, { IUsersAvatarsWallProps } from '../UsersAvatarsWall';

const props: IUsersAvatarsWallProps = {
  users: Array(10).fill(githubUser),
  onAvatarClick: jest.fn(),
};

describe('UsersAvatarsWall component', () => {
  const component = shallow(<UsersAvatarsWall {...props} />);
  it('should match snapshot', () => {
    expect(component).toMatchSnapshot();
  });

  const avatars = component.find('.users-avatars-wall__avatar');
  it('should display all users', () => {
    expect(avatars.length).toEqual(props.users.length);
  });

  it('should call onAvatarClick on avatar click', () => {
    avatars.at(0).simulate('click');
    expect(props.onAvatarClick).toHaveBeenCalledTimes(1);
  });
});
