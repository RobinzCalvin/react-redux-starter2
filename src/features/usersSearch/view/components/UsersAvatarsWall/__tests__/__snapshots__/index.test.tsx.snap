// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UsersAvatarsWall component should match snapshot 1`] = `
<div
  className="users-avatars-wall"
>
  <ul
    className={
      Object {
        "has": [Function],
        "is": [Function],
        "mix": [Function],
        "split": [Function],
        "state": [Function],
        "toString": [Function],
      }
    }
  >
    <Preloader
      backgroundColor="#fff"
      isShown={false}
      size={100}
    />
    <li
      className={
        Object {
          "has": [Function],
          "is": [Function],
          "mix": [Function],
          "split": [Function],
          "state": [Function],
          "toString": [Function],
        }
      }
      key="1"
      onClick={[Function]}
    >
      <img
        className={
          Object {
            "has": [Function],
            "is": [Function],
            "mix": [Function],
            "split": [Function],
            "state": [Function],
            "toString": [Function],
          }
        }
        onLoad={[Function]}
        src="http://the-url.com&s=70"
      />
    </li>
    <li
      className={
        Object {
          "has": [Function],
          "is": [Function],
          "mix": [Function],
          "split": [Function],
          "state": [Function],
          "toString": [Function],
        }
      }
      key="1"
      onClick={[Function]}
    >
      <img
        className={
          Object {
            "has": [Function],
            "is": [Function],
            "mix": [Function],
            "split": [Function],
            "state": [Function],
            "toString": [Function],
          }
        }
        onLoad={[Function]}
        src="http://the-url.com&s=70"
      />
    </li>
    <li
      className={
        Object {
          "has": [Function],
          "is": [Function],
          "mix": [Function],
          "split": [Function],
          "state": [Function],
          "toString": [Function],
        }
      }
      key="1"
      onClick={[Function]}
    >
      <img
        className={
          Object {
            "has": [Function],
            "is": [Function],
            "mix": [Function],
            "split": [Function],
            "state": [Function],
            "toString": [Function],
          }
        }
        onLoad={[Function]}
        src="http://the-url.com&s=70"
      />
    </li>
    <li
      className={
        Object {
          "has": [Function],
          "is": [Function],
          "mix": [Function],
          "split": [Function],
          "state": [Function],
          "toString": [Function],
        }
      }
      key="1"
      onClick={[Function]}
    >
      <img
        className={
          Object {
            "has": [Function],
            "is": [Function],
            "mix": [Function],
            "split": [Function],
            "state": [Function],
            "toString": [Function],
          }
        }
        onLoad={[Function]}
        src="http://the-url.com&s=70"
      />
    </li>
    <li
      className={
        Object {
          "has": [Function],
          "is": [Function],
          "mix": [Function],
          "split": [Function],
          "state": [Function],
          "toString": [Function],
        }
      }
      key="1"
      onClick={[Function]}
    >
      <img
        className={
          Object {
            "has": [Function],
            "is": [Function],
            "mix": [Function],
            "split": [Function],
            "state": [Function],
            "toString": [Function],
          }
        }
        onLoad={[Function]}
        src="http://the-url.com&s=70"
      />
    </li>
    <li
      className={
        Object {
          "has": [Function],
          "is": [Function],
          "mix": [Function],
          "split": [Function],
          "state": [Function],
          "toString": [Function],
        }
      }
      key="1"
      onClick={[Function]}
    >
      <img
        className={
          Object {
            "has": [Function],
            "is": [Function],
            "mix": [Function],
            "split": [Function],
            "state": [Function],
            "toString": [Function],
          }
        }
        onLoad={[Function]}
        src="http://the-url.com&s=70"
      />
    </li>
    <li
      className={
        Object {
          "has": [Function],
          "is": [Function],
          "mix": [Function],
          "split": [Function],
          "state": [Function],
          "toString": [Function],
        }
      }
      key="1"
      onClick={[Function]}
    >
      <img
        className={
          Object {
            "has": [Function],
            "is": [Function],
            "mix": [Function],
            "split": [Function],
            "state": [Function],
            "toString": [Function],
          }
        }
        onLoad={[Function]}
        src="http://the-url.com&s=70"
      />
    </li>
    <li
      className={
        Object {
          "has": [Function],
          "is": [Function],
          "mix": [Function],
          "split": [Function],
          "state": [Function],
          "toString": [Function],
        }
      }
      key="1"
      onClick={[Function]}
    >
      <img
        className={
          Object {
            "has": [Function],
            "is": [Function],
            "mix": [Function],
            "split": [Function],
            "state": [Function],
            "toString": [Function],
          }
        }
        onLoad={[Function]}
        src="http://the-url.com&s=70"
      />
    </li>
    <li
      className={
        Object {
          "has": [Function],
          "is": [Function],
          "mix": [Function],
          "split": [Function],
          "state": [Function],
          "toString": [Function],
        }
      }
      key="1"
      onClick={[Function]}
    >
      <img
        className={
          Object {
            "has": [Function],
            "is": [Function],
            "mix": [Function],
            "split": [Function],
            "state": [Function],
            "toString": [Function],
          }
        }
        onLoad={[Function]}
        src="http://the-url.com&s=70"
      />
    </li>
    <li
      className={
        Object {
          "has": [Function],
          "is": [Function],
          "mix": [Function],
          "split": [Function],
          "state": [Function],
          "toString": [Function],
        }
      }
      key="1"
      onClick={[Function]}
    >
      <img
        className={
          Object {
            "has": [Function],
            "is": [Function],
            "mix": [Function],
            "split": [Function],
            "state": [Function],
            "toString": [Function],
          }
        }
        onLoad={[Function]}
        src="http://the-url.com&s=70"
      />
    </li>
  </ul>
</div>
`;
